cmd/identity/batch.go

```diff
@@ -0,0 +1,200 @@
+		MinDifficulty int    `help:"minimum difficulty to output" default:"30"`
+		Concurrency   int    `help:"worker concurrency" default:"4"`
+		OutputDir     string `help:"output directory to place keys" default:"."`
+	}
+)
+
+func init() {
+	rootCmd.AddCommand(keyGenerateCmd)
+	cfgstruct.Bind(keyGenerateCmd.Flags(), &keyCfg)
+}
+
+func cmdKeyGenerate(cmd *cobra.Command, args []string) (err error) {
+	ctx, cancel := context.WithCancel(process.Ctx(cmd))
+	defer cancel()
+
+	err = os.MkdirAll(keyCfg.OutputDir, 0700)
+	if err != nil {
+		return err
+	}
+
+	var group errgroup.Group
+	defer func() {
+		err = errs.Combine(err, group.Wait())
+	}()
+
+	screen, err := cui.NewScreen()
+	if err != nil {
+		return err
+	}
+	group.Go(func() error {
+		defer cancel()
+		err := screen.Run()
+		return errs.Combine(err, screen.Close())
+	})
+
+	counter := new(uint32)
+	diffCounts := [256]uint32{}
+	return identity.GenerateKeys(ctx, uint16(keyCfg.MinDifficulty), keyCfg.Concurrency,
+		func(k *ecdsa.PrivateKey, id storj.NodeID) (done bool, err error) {
+			difficulty, err := id.Difficulty()
+			if err != nil {
+				return false, err
+			}
+
+			if int(difficulty) > len(diffCounts) {
+				atomic.AddUint32(&diffCounts[len(diffCounts)-1], 1)
+			} else {
+				atomic.AddUint32(&diffCounts[difficulty], 1)
+			}
+
+			if err := renderStats(screen, diffCounts[:]); err != nil {
+				return false, err
+			}
+
+			genName := fmt.Sprintf("gen-%02d-%d", difficulty, atomic.AddUint32(counter, 1))
+			err = saveIdentityTar(filepath.Join(keyCfg.OutputDir, genName), k, id)
+			return false, err
+		})
+}
+
+func saveIdentityTar(path string, key *ecdsa.PrivateKey, id storj.NodeID) error {
+	ct, err := peertls.CATemplate()
+	if err != nil {
+		return err
+	}
+
+	caCert, err := peertls.NewCert(key, nil, ct, nil)
+	if err != nil {
+		return err
+	}
+
+	ca := &identity.FullCertificateAuthority{
+		Cert: caCert,
+		ID:   id,
+		Key:  key,
+	}
+
+	ident, err := ca.NewIdentity()
+	if err != nil {
+		return err
+	}
+
+	tarData := new(bytes.Buffer)
+	tw := tar.NewWriter(tarData)
+
+	caCertBytes, caCertErr := peertls.ChainBytes(ca.Cert)
+	caKeyBytes, caKeyErr := peertls.KeyBytes(ca.Key)
+	identCertBytes, identCertErr := peertls.ChainBytes(ident.Leaf, ident.CA)
+	identKeyBytes, identKeyErr := peertls.KeyBytes(ident.Key)
+	if err := errs.Combine(caCertErr, caKeyErr, identCertErr, identKeyErr); err != nil {
+		return err
+	}
+
+	if err := errs.Combine(
+		writeToTar(tw, "ca.cert", caCertBytes),
+		writeToTar(tw, "ca.key", caKeyBytes),
+		writeToTar(tw, "identity.cert", identCertBytes),
+		writeToTar(tw, "identity.key", identKeyBytes),
+	); err != nil {
+		return err
+	}
+
+	if err := tw.Close(); err != nil {
+		return errs.Wrap(err)
+	}
+
+	if err = ioutil.WriteFile(path+".tar", tarData.Bytes(), 0600); err != nil {
+		return errs.Wrap(err)
+	}
+	return nil
+}
+
+func writeToTar(tw *tar.Writer, name string, data []byte) error {
+	err := tw.WriteHeader(&tar.Header{
+		Name: name,
+		Mode: 0600,
```

Comment:
I think certs are supposed to be 0644 based on  https://github.com/storj/storj/blob/master/pkg/identity/identity_test.go#L126